spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindableProxyFactory.java

```diff
@@ -123,50 +122,49 @@ public void afterPropertiesSet() throws Exception {
-				try {
-					Input input = AnnotationUtils.findAnnotation(method, Input.class);
-					if (input != null) {
-						String name = BindingBeanDefinitionRegistryUtils.getChannelName(input, method);
-						MessageChannel sharedChannel = locateSharedChannel(name);
-						if (sharedChannel == null) {
-							inputHolders.put(name, new ChannelHolder(
-									channelFactory.createChannel(name, method.getReturnType()), true));
-						}
-						else {
-							configureSharedMessageChannel(name, method.getReturnType(), sharedChannel);
+				Input input = AnnotationUtils.findAnnotation(method, Input.class);
+				if (input != null) {
+					String name = BindingBeanDefinitionRegistryUtils.getChannelName(input, method);
+					Class<? extends MessageChannel> channelType = (Class<? extends MessageChannel>) method.getReturnType();
+					MessageChannel sharedChannel = locateSharedChannel(name);
+					if (sharedChannel == null) {
+						inputHolders.put(name, new ChannelHolder(createBindableChannel(name, channelType), true));
+					}
+					else {
+						if (!channelType.isAssignableFrom(sharedChannel.getClass())) {
+							bridgeSharedChannel(channelType, sharedChannel);
 						}
 					}
 				}
-				catch (Exception e) {
-					throw new RuntimeException(e);
-				}
 			}
 		});
 		ReflectionUtils.doWithMethods(type, new ReflectionUtils.MethodCallback() {
 			@Override
 			public void doWith(Method method) throws IllegalArgumentException {
-				try {
-					Output output = AnnotationUtils.findAnnotation(method, Output.class);
-					if (output != null) {
-						String name = BindingBeanDefinitionRegistryUtils.getChannelName(output, method);
-						MessageChannel sharedChannel = locateSharedChannel(name);
-						if (sharedChannel == null) {
-							outputHolders.put(name, new ChannelHolder(
-									channelFactory.createChannel(name, method.getReturnType()), true));
-						}
-						else {
-							configureSharedMessageChannel(name, method.getReturnType(), sharedChannel);
+				Output output = AnnotationUtils.findAnnotation(method, Output.class);
+				if (output != null) {
+					String name = BindingBeanDefinitionRegistryUtils.getChannelName(output, method);
+					Class<? extends MessageChannel> channelType = (Class<? extends MessageChannel>) method.getReturnType();
+					MessageChannel sharedChannel = locateSharedChannel(name);
+					if (sharedChannel == null) {
+						outputHolders.put(name, new ChannelHolder(createBindableChannel(name, channelType), true));
+					}
+					else {
+						if (!channelType.isAssignableFrom(sharedChannel.getClass())) {
+							bridgeSharedChannel(channelType, sharedChannel);
 						}
 					}
 				}
-				catch (Exception e) {
-					throw new RuntimeException(e);
-				}
 			}
 
 		});
 	}
 
+	private MessageChannel createBindableChannel(String name, Class<? extends MessageChannel> channelType) {
+			return isPollable(channelType) ? this.channelFactory.createPollableChannel(name) :
```

Comment:
Sorry. I always thought my IDE would fix it when I save but apparent it didn't do so :(. Pushed the fix.